proposals/gc/Overview.md

```diff
@@ -231,7 +242,76 @@ An alternative is to provide [primitive support](#closures) for closures, e.g. a
 
 ### Parametric Polymorphism
 
-TODO: via type `anyref` and `intref`
+* Dynamic languages or static languages with sufficiently expressive parametric polymorphism (generics) often require a *uniform representation*, where all its data types are represented in a single word.
+* Typically, pointer tagging is used to unbox small scalars.
+* Want to be able to represent this with type `anyref`.
+
+Contrived example (fictional language):
+```
+function make_pair<A, B>(a : A, b : B) : (A, B) {
+  return (a, b);
+}
+
+class C {...};
+function f() {
+  make_pair<Bool, Bool>(true, false);
+  make_pair<C, C>(new C, new C);
+}
+
+function fst<A>(p : (A, A)) : A { let (a, _) = p; return a }
+function snd<A>(p : (A, A)) : A { let (_, a) = p; return a }
+
+function g(p1 : (Bool, Bool), p2 : (C, C), pick : <A> ((A, A)) -> A) : C {
+  if (pick<Bool>(p1))
+    return pick<C>(p2);
+  else
+    return new C;
+}
+```
+
+Here, `make_pair` as well as `fst` and `snd` need to be able to operate on any type of pair. Furthermore, `fst` and `snd` cannot simply be type-specialised at compile time, because that would be insufficient to compile `g`, which takes a polymorphic function as an argument and instantiates it with multiple different types. Such *first-class* polymorphism is not expressible with compile time techniques such as C++ templates, but common-place in many languages (including OO ones like Java or C#, where it can be emulated via generic methods). Untyped languages like JavaScript or Scheme trivially allow such programs as well.
```

Comment:
I think this could be explained more precisely. In particular, there will be programming languages for which the type checker is able to determine the return type of the use of `pick` in `g`, and thus can compile it to safe code that does not need a `cast_down` (e.g. when translated directly to LLVM, C or x86). So the need for a RTT / `cast_down` is not just about first class polymorphism, but it is about a mismatch between the language and what wasm types can express (since we apparently don't want to have the concept of statically typed generics at the wasm level). This a tradeoff we're choosing, i.e. we could instead decide to make wasm types a whole lot more complicated in exchange for needing less or no `cast_down` ops or need for a RTT field in all "objects".

This is similar to what Java does, which I am not a fan of, though I understand why dealing with generics at the wasm level may not be impractical. On a related note, letting inefficiencies like this creep into the format we're encouraging engines to do their own type analysis to omit `cast_down` and RTT fields where possible (it wouldn't surprise me if Hotspot does this already).

Reply:
@aardappel, I agree that this is a choice we make. I added a couple more sentences mentioning the possible future addition of type parameters to Wasm to avoid this overhead.

To be clear, the problem isn't the ability to deduce the return type -- AFAICT, all languages are able to do that (that's the point of generics after all). The problem is that the compilation of `fst` and `snd` itself must not depend on that type because you generally cannot tell, at compile time, the set of all types they will be instantiated with (static analysis can do that in many cases but not all). So compile-time type specialisation does not generally work with first-class polymorphism. Unless you are willing to implement runtime compilation and type specialisation (like C# / .NET) you have to rely on a type-agnostic compilation scheme, and short of having type parameters in Wasm, going to the top type and casting back is the only alternative available.

Added a version of this as explanation.

Reply:
Ah, good point about `fst` and `snd`, thanks. Good to hear you think type parameters could still be added in the future.